types/react/index.d.ts

```diff
@@ -418,6 +418,31 @@ declare namespace React {
         refs: {
             [key: string]: ReactInstance
         };
+
+        // Static properties copied from ComponentClass interface.
+        /**
+         * The validity of the type of `propTypes` cannot be checked
+         * automatically at the definition of a component class because the
+         * required type depends on `P`.  We recommend checking it using
+         * `asPropTypes` from the `react-typescript-helpers` package.  If you
```

Comment:
What is this package 🤔 

https://www.npmjs.com/package/react-typerscript-helpers 404s

Reply:
Check your spelling: https://www.npmjs.com/package/react-typescript-helpers

Reply:
Oops 🙈

Reply:
The example code in the README doesn't look valid 🤔

Reply:
The actual code doesn't look valid either, or at least with this PR, your constraint in the returned functions only work by accident because of the types in `FunctionComponent` (or worse, widen to `any`).

Reply:
> The example code in the README doesn't look valid (`ReferenceError`) :thinking:

I knew I was going to make some mistake.  I fixed the occurrences of `MyComponent` that should have been `MyComponentClass` (in master, not yet published to npm).  Did you find anything else wrong?

Reply:
> The actual code doesn't look valid either, or at least with this PR, your constraint in the returned functions only work by accident because of the types in `FunctionComponent` (or worse, widen to `any`).

Sorry, I don't understand what you mean.  When I modify `package.json` to point to `@types/react` with this PR, the test cases in `test.tsx` still compile successfully, and if I deliberately put the wrong type in `defaultProps`, I get the expected compile error.

Reply:
```ts
export function asDefaultProps<P>(component: LooseComponentType<P>) {
    return <DP extends React.ComponentType<P>["defaultProps"]>
        (defaultProps: DP) => defaultProps;
}
```

`ComponentType<P>["defaultProps"]` is going to be `ComponentClass<P>["defaultProps"] | FunctionComponent<P>["defaultProps"]`, which will expand to `object | Partial<P>`. This will still work right now because we're not dealing with `FunctionComponent`'s problem, but it's more of a case of working by accident than working as intended.

Probably needs `<DP extends Partial<P>>`.

A similar problem is present with the `asPropTypes` function.

Reply:
> `ComponentType<P>["defaultProps"]` is going to be `ComponentClass<P>["defaultProps"] | FunctionComponent<P>["defaultProps"]`, which will expand to `object | Partial<P>`.

No, this PR does not change `ComponentClass<P>["defaultProps"]`; it is still `Partial<P>`.  The `object` is only in `Component.defaultProps`.  Do you foresee changing the type of `ComponentClass.defaultProps` or `FunctionComponent.defaultProps`?

Reply:
Ah, I'm thinking of https://github.com/DefinitelyTyped/DefinitelyTyped/issues/30791 and what it would take to fix it.

It's still probably better to be future-proof, either by exporting `type` aliases from `react` or by directly using `Partial` / `ValidationMap` in the `extends` clauses.

Reply:
Done in master of react-typescript-helpers.  (I'll wait to publish to npm until the review of this pull request is done.)